### PR TITLE
Service handler based libkbfs.KeybaseService impl instead of loopback RPC

### PR DESCRIPTION
**Pull request by gabriel**
This is still in progress but the idea is, instead of reading/writing to a socket file that the service is listening to and doing RPC socket msgpack stuff, you can call the engine directly.

KBFS has a very clear interface/definition of what it needs from the service. We can implement it directly by calling into the engine stuff.

### CODE REVIEW

File: go/engine/resolve_identify2.go [hunk 1]
```diff
@@ -134,3 +134,12 @@ func (e *ResolveThenIdentify2) TrackToken() keybase1.TrackToken {
 func (e *ResolveThenIdentify2) ConfirmResult() keybase1.ConfirmResult {
 	return e.i2eng.ConfirmResult()
 }
+
+func Resolve(g *libkb.GlobalContext, assertion string) (libkb.NormalizedUsername, keybase1.UID, error) {
```
Comment (maxtaco) [inline go/engine/resolve_identify2.go]: `ResolveRun`?

File: go/bind/service.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	// TODO: No caching is not enabled yet
+}
+
+func (s *kbservice) FlushUserUnverifiedKeysFromLocalCache(ctx context.Context, uid keybase1.UID) {
+	// TODO: No caching is not enabled yet
+}
+
+func (s *kbservice) Shutdown() {
+	// There a no resources to cleanup (this is useful for RPC impl)
+}
```
Comment (maxtaco) [inline go/bind/service.go]: What about the `crypto` RPCs?

**Comment by maxtaco** [comment]
Seems very straightforward, I like this approach! How much more work is left?

**Comment by maxtaco** [comment]
Also, how would the KBFS code need to change (if at all?)

File: go/bind/service.go (hunk 2 above)
Comment (gabriel) [inline go/bind/service.go]: Oops, didn't see those.. we can do the same thing...

I opened a PR for some changes so we can pass an engine based CryptoClient here: https://github.com/keybase/kbfs/pull/215

There are only a few crypto methods (sign+unbox) and we will be forwarding into engine calls like the service.

**Comment by gabriel** [comment]
Not too much work left... need some small KBFS changes mentioned above... 

Do we need to implement caching in these impls, or the engine caches, right? We flush methods can be no-ops in this engine based approach?

**Comment by gabriel** [comment]
With the KBFS changes, the crypto client impl is super simple (forwards to engine):
https://github.com/keybase/client/pull/3723/commits/796b4403b5eb8053494b869ed38468c7c080ada3

**Comment by gabriel** [comment]
Ok, I did a FS list of a private folder (hooked up a button in search tab to trigger it) and was able to look at my files in `/keybase/private/gabrielh`.

I had to add a temporary (only for debug) auto confirming `IdentifyUI` impl for mobile, until we implement that for real.

Just wanted to re-iterate that this doesn't bypass the protocol, this PR is mostly code thats connecting/configuring stuff to bypass the connection/serialization layer and straight into the engine but it still uses the protocol stuff of course.

For example, this line:

```go
srv.Register(keybase1.FsProtocol(fsrpc.NewFS(kbfsConfig, kbCtx.Log)))
```

This is telling the service to use the FsProtocol in keybase/kbfs/fsrpc instead of the RPC one (use it directly). The RPC one would eventually get to this same class but would have to go through serialization and connection to do it, which is a little awkward because we are go talking to go in the same runtime.

**Comment by gabriel** [comment]
So there is behavior in KBFS' `libkbfs.KeybaseDaemonRPC` that we want, and not RPC specific, so I made a PR to separate that out: https://github.com/keybase/kbfs/pull/220

Then we don't need to wrap a lot of the service methods, and get behavior that KBFS expects from a KeybaseService implementation.

**Comment by gabriel** [comment]
Also to clarify:

By RPC I mean connection based RPC behavior. The protocol is the source of truth, is the API, is a layer we shouldn't bypass.

The allows us to go from:

```
 go -> protocol -> msgpack -> connection -> msgpack ->  protocol -> go 
```

to:

```
go -> protocol -> go
```

**Comment by maxtaco** [comment]
I really appreciate the work here, it isn't much fun to refactor code in lots of places to do this type of integration. 

I have a lot of thoughts on this. First, what is the current state of the app? With my fix from yesterday does everything works as we intend (with regards to this issue)?

Second, back in the early days of the go app we implemented two different code paths for standalone and daemon modes. There was a lot of shared code but also a lot of annoying glue. Anytime we added something new there was glue we had to add and it slowed us down.  After a while we switched **to** the current loopback system and stripped out about a thousand lines of code and also were able to develop faster. This PR adds about 300 LOC of glue and refactors and new levels of indirection.  We will have to add more whenever we add new features. The levels of indirection can make it harder to understand the code. 

Third, the only argument is my mind against the loopback system is performance.  I am really not swayed by the argument that we are doing unnecessary translations because it doesn't affect the user or the developer.  But back to the performance question, if there is evidence that we have a bottleneck introduced through loopback then it is a compelling argument to adopt this approach. I am skeptical though since the connection between the service and kbfs is pretty low bandwidth. 

Fourth, I didn't see any support for notifications that the service sends kbfs.  There are several of these, which we send on use changed or keys changed.  I could have missed it.

**Comment by gabriel** [comment]
OK, I explored this a bit more and yeah, calling into the engine wasn't the right approach.

I took another pass and I think I have some (minor, I think) refactoring which would be good to have (even if we don't want to bypass rpc). So instead of calling directly into the engine, this PR uses the same protocol impls the service is using, the handlers.

The service handlers are constructed from RPC connections, but if we split them into non-RPC and RPC types then we can use them in the bind package without RPC.

I did this to `UserHandler`, `SessionHandler`, `IdentifyHandler`, `FavoriteHandler` and `KBFSHandler` which are the 5 protocols currently used by KBFS.

It s a pretty minor change, and yeah it does add a level of abstraction, but I think is really good for a couple reasons:
- Good to be have protocol implemented in semi-agnostic way (if we swapped msgpack for something else for example)
- Separation of concerns (loading a user protocol impl, shouldn't require a connection or rpc package), less spaghetti

So I made these changes in my exploration and tested it and it works well... This PR doesn't flip the switch on non-RPC behavior, the old behavior is still enabled. It does add a non-RPC impl:

```go
func (s serviceCn) NewKeybaseService(config libkbfs.Config, params libkbfs.InitParams, ctx libkbfs.Context, log logger.Logger) (libkbfs.KeybaseService, error) {
	// To use a local non-RPC service provider:
	// return newServiceProvider(s.ctx, config, log), nil

	// To use an RPC based service provider:
	return newServiceRPCProvider(s.ctx, config, log), nil
}

func (s serviceCn) NewCrypto(config libkbfs.Config, params libkbfs.InitParams, ctx libkbfs.Context, log logger.Logger) (libkbfs.Crypto, error) {
	// To use a local non-RPC crypto provider:
	// return newCrypto(s.ctx, config, log), nil

	// To use an RPC based crypto provider:
	return newCryptoRPC(s.ctx, config, log), nil
}
```

So we can swap from RPC to non-RPC. AFAIK, there is no code duplication and the only thing being bypassed is the connection/msgpack layer, and all other behavior is the same.

Ok so to run through your points, which are all good:

(1) This new PR (with non-RPC version enabled) should work exactly the same as RPC.
(2) I think with the new changes (using service handlers), we aren't duplicating any code paths.
(3) Performance isn't the only thing I have against loopback. In fact, my reasoning for this isn't really performance, it's more all the additional complexity, the requirement of msgpack, rpc and other dependencies in general, and all the unnecessary behavior associated with that. My argument is more about good abstractions.
(4) This KBFS PR: https://github.com/keybase/kbfs/pull/220 separates RPC and non-RPC behavior so caching and notifications should be exactly same now
(5) I suppose we do loose RPC debugging but maybe we should add this into the protocol layer?

If you still like the RPC approach, then could we maybe keep this PR to have the non-RPC version (should only be 20+ lines of code), so that we have the option in the future? And also I think an important consideration to keep RPC and non-RPC behavior separate for good code/package quality?

Also I want to say up front, I kind of agree with go about bad abstractions being worse than duplicated code, but I think this is a good abstraction and little/no duplicated code, so? :)

LMK if we should google hangout...

(Also this won't pass CI yet until KBFS PR is merged.)

**Comment by gabriel** [comment]
Re (5): A benefit of RPC, was that you could enable logging and get logging for all the calls from service to KBFS automatically. 

- It could be that, we already have a lot of logging around all these calls and having additional logging here isn't that much of a benefit (we might not miss it)
- If we need to debug something we can add logging as needed
- We could have our protocol go code generate logging proxies for all the interfaces (since there is no way to do this dynamically in go)

@cjb what do you think?

**Pull request closed without merging**